### PR TITLE
EDACC dataset automatic speech recognition

### PR DESCRIPTION
**Pull request by uwanny**
## What?

Add an ASR recipe for EDACC dataset, an accent English speech dataset ([website](https://groups.inf.ed.ac.uk/edacc/)) trained using WavLM + transformer.

## Why?

There are little accented English speech corpus in the ESPnet framework.

### Specification:
1. The provided data has only dev set and test set, so part of the dev set was set as train set in data prep stage, the specific number of utterances for train set could be set in [data_prep.py](https://github.com/espnet/espnet/pull/5996/files#diff-f60539c1f7698c08d8d4acdafadac2693e49f95c814ec2cb1223dcedc8520c46).
2. For the convenience of dump data in stage 3, to avoid OOM error because of loading too large wav file, too long wav file was split in [data.sh](https://github.com/espnet/espnet/pull/5996/files#diff-4051cf648b046599122a684ab4fa0c54ecb076f94ce313a7986d4345983064f4)
3. For the convenience of decoding in stage 11 in test set, to avoid OOM becasue of too long utterance, it is possible to set specific utterance and alignment rules (you can get from CTC segmentation tools)  in [truncate_test.py](https://github.com/espnet/espnet/pull/5996/files#diff-171e86a9abfcc0a08732a357b838ee7f6ae3dbd130d21a1ec98aecbd90ae35cb) to split the test utterance.

### Problem:
1. Training using wavLM + transformer could work as the training stage actually helps, and the WER of test set did decrease, the training accuracy did increase during the training. The WER is 51 for test set which is larger than paper reported WER (18.7-36.1), however, the paper used some well-established pretrained model for inference, making the result better. I planned to make the PR for the current stage and tried to find some better method if have time. 

### Reference:
[1] Sanabria, R., Bogoychev, N., Markl, N., Carmantini, A., Klejch, O., & Bell, P. (2023). The Edinburgh International Accents of English Corpus: Towards the Democratization of English ASR. In ICASSP 2023.

### CODE REVIEW

**Comment by sw005320** [comment]
Can you add some discussions about how you use the dev set (you split it to train and valid, right?)?
Also, please describe your treatments for the long segments.

File: egs2/edacc/asr1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    python3 local/data_prep.py "${EDACC}/edacc_v1.0" "data" "${output_dir}"
+
+    # # split the too long test utterance used for decoding section if necessary,
+    # # the alignment is based on CTC segmentation tool
+    # python3 local/truncate_test.py "data/test"
+
+    # # make training data from dev, as original data has no training data
+    # utils/subset_data_dir.sh --first data/dev 5000 "data/${train_set}"
+    # n=$(($(wc -l < data/dev/segments) - 5000))
+    # utils/subset_data_dir.sh --last data/dev ${n} "data/${valid_set}"
```
Comment (ftshijt) [inline egs2/edacc/asr1/local/data.sh:81]: As you mentioned in the PR, could you please explain a bit more on how you process the ctc segmentation? As I get into truncate test, I did not see any usage of CTC segmentation. It is still a bit unclear to me on how you fix the problem with that.

Additionally, according to your code, it seems that only EDACC-C30 is split?

Comment (ftshijt) [reply egs2/edacc/asr1/local/data.sh:81]: Since the process seems complicated, how about spliting the second stage into multiple smaller ones? We introduce stages, mostly for this purpose.

File: egs2/edacc/asr1/local/data_prep.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        3
+    ]  # should be "downloads/edacc_v1.0/data/segmentation"
+    generate_data_files(
+        target_root,
+        edacc_root,
+        segmented_audio_path,
+        segment_length=1883,
+        utter_number=5000,
+    )
+    # print("data prep script completed successfully.")
```
Comment (ftshijt) [inline egs2/edacc/asr1/local/data_prep.py:149]: You may safely remove the line if not used anymore in the final version.

File: egs2/edacc/asr1/run.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+set -e
+set -u
+set -o pipefail
+
+
+
+test_set="test test_sub"
+train_set="dev_train"
+valid_set="dev_non_train"
+nbpe=3884 # 3884 vocabulary size of bpe could cover all the sentence in the edacc dataset
```
Comment (ftshijt) [inline egs2/edacc/asr1/run.sh:13]: The number seems weird to me. Could you please elaborate a bit on why 3884 is selected?

File: egs2/edacc/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+asr_config=conf/train_asr_wavlm_transformer.yaml
+# asr_config=conf/train_asr_transformer_finetune.yaml
+inference_config=conf/decode_asr.yaml
+
+./asr.sh \
+    --nj 32 \
+    --inference_nj 2 \
+    --gpu_inference true \
```
Comment (ftshijt) [inline egs2/edacc/asr1/run.sh:21]: Leave these to default would be fine (by removing them)

File: egs2/edacc/asr1/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+- pytorch version: `pytorch 2.0.1`
+- Git hash: `70d2c9cd76ea066fe5c84d148799b6c94e58f57c`
+  - Commit date: `Thu Dec 12 00:28:13 2024 -0500`
+
+## exp/asr_train_asr_wavlm_transformer_raw_en_bpe3884_sp
+### WER
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|decode_asr_asr_model_valid.acc.ave/test|9300|163389|56.1|31.3|12.6|7.6|51.5|87.3|
```
Comment (ftshijt) [inline egs2/edacc/asr1/README.md:22]: The WER seems much worse than what the paper stated. Is there a possible reason for that?

**Review by ftshijt — COMMENTED** [review]
Thanks! Please also indicate your entry at egs2/README.md for the new data

File: egs2/edacc/asr1/local/data.sh (hunk 1 above)
Comment (uwanny) [reply egs2/edacc/asr1/local/data.sh:81]: > As you mentioned in the PR, could you please explain a bit more on how you process the ctc segmentation? As I get into truncate test, I did not see any usage of CTC segmentation. It is still a bit unclear to me on how you fix the problem with that.
> 
> Additionally, according to your code, it seems that only EDACC-C30 is split?

1. Yes, only EDACC-C30 is split, this is especially used for RM-shared in default memory case, to avoid OOM.
2. In truncate test, you should customize your segmentation rules, such as what text you need, start time and end time. And these parameters should be obtained in advance from a segmentation tool, I tested using segmentaion tool in my side, I didn't include actual text here because it will cause line too long error

Comment (uwanny) [reply egs2/edacc/asr1/local/data.sh:81]: > Since the process seems complicated, how about spliting the second stage into multiple smaller ones? We introduce stages, mostly for this purpose.

Yes, I will do that

File: egs2/edacc/asr1/run.sh (hunk 3 above)
Comment (uwanny) [reply egs2/edacc/asr1/run.sh:13]: Because it seems like in stage 5, when trying to use 5000 vocab size, it indicates that the dataset does not have sufficient diversity to support a vocabulary size of 5000, and the maximum feasible vocabulary size for this dataset is 3884.

File: egs2/edacc/asr1/README.md (hunk 5 above)
Comment (uwanny) [reply egs2/edacc/asr1/README.md:22]: The Reason is [here](https://github.com/espnet/espnet/pull/5996#issue-2759070320)

File: egs2/edacc/asr1/run.sh (hunk 3 above)
Comment (ftshijt) [reply egs2/edacc/asr1/run.sh:13]: usually, we use a smaller vocab size as the bpe is for word pieces. 

If you are using 3884, it actually returns to a word-based model (this could be one major cause of the poor WER you got from the model, as the word is very sparse with a small dataset).

Given that the word size is small, I would suggest you to go with a smaller vocab size (e.g., 500 or even 100). I expect it could make a significant improvement to the model's performance.

File: egs2/edacc/asr1/local/data.sh (hunk 1 above)
Comment (ftshijt) [reply egs2/edacc/asr1/local/data.sh:81]: I see, in that case, it would be helpful if you could make it clear in the script that this step is optional.

For a better understanding and future usage of the script you provided, you might consider putting an example in the readme with markdown style (won't trigger the length warning).

**Comment by ftshijt** [comment]
As I mentioned in the detailed discussion, I feel the current performance is largely degraded due to the large vocab size. It is very likely that the system can get better if you can use less word size for training.

File: egs2/edacc/asr1/local/data.sh (hunk 1 above)
Comment (uwanny) [reply egs2/edacc/asr1/local/data.sh:81]: Good advice, will do

**Comment by uwanny** [comment]
> As I mentioned in the detailed discussion, I feel the current performance is largely degraded due to the large vocab size. It is very likely that the system can get better if you can use less word size for training.

I will try to retrain it, thank you

**Comment by uwanny** [comment]
By trying the BPE to 500 and make some ablation exps, the training stage of models is a little better than previous, as shown in the accuracy figure (using larger BPE can only achieved smaller accuracy) 
image
The WER curve shows a big vibration:
image

However, the actual WER in test set could not show a big improvement (comparable with article reported), I think this is becasue firstly in article they decode the sentence every 30s, but they didn't provide the actual transcription in the given for decoding in that way, which means the provided dataset has test set split in different ways as they did in the paper. Secondly, even it is not quite meaningful by comparing the result directly with paper, the current WER in test set, I think is still a bit higher compared with some baseline result in ASR, but actually by trying several parameters the best test result are the one using BPE 500 with training curve shown in the previous two figures (decrease the BPE to 100 could increase the training accuracy a little more, but can't guarantee a improved performance in test case based on my test), this model could achieve a result of:

### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_asr_model_valid.acc.ave/test|9300|163389|59.5|29.1|11.4|6.7|47.2|88.3|

### CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_asr_model_valid.acc.ave/test|9300|792343|76.0|11.0|13.0|8.3|32.3|88.3|

### TER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_asr_asr_model_valid.acc.ave/test|9300|268438|62.1|26.2|11.6|12.2|50.1|88.3|

Better than the previous:

### WER

| dataset                                     | Snt  | Wrd    | Corr | Sub  | Del  | Ins | Err  | S.Err |
| ------------------------------------------- | ---- | ------ | ---- | ---- | ---- | --- | ---- | ----- |
| decode_asr_asr_model_valid.acc.ave/test     | 9300 | 163389 | 56.1 | 31.3 | 12.6 | 7.6 | 51.5 | 87.3  |


### CER

| dataset                                     | Snt  | Wrd    | Corr | Sub  | Del  | Ins | Err  | S.Err |
| ------------------------------------------- | ---- | ------ | ---- | ---- | ---- | --- | ---- | ----- |
| decode_asr_asr_model_valid.acc.ave/test     | 9300 | 792343 | 68.9 | 13.5 | 17.7 | 8.3 | 39.4 | 87.3  |


### TER

| dataset                                     | Snt  | Wrd    | Corr | Sub  | Del  | Ins | Err  | S.Err |
| ------------------------------------------- | ---- | ------ | ---- | ---- | ---- | --- | ---- | ----- |
| decode_asr_asr_model_valid.acc.ave/test     | 9300 | 206542 | 54.1 | 29.0 | 16.9 | 9.2 | 55.1 | 87.3  |

Based on the previous comment and discussion, I think the model achieved best result given current circumstances, the training config used in the best model is: Please give some advice if you have, otherwise I will update using my current result.
```
freeze_param: [
"frontend.upstream"
]

frontend: s3prl
frontend_conf:
    frontend_conf:
        upstream: wavlm_base_plus
    download_dir: ./hub
    multilayer_feature: True

preencoder: linear
preencoder_conf:
   input_size: 768  # Note: If the upstream is changed, please change this value accordingly.
   output_size: 80

encoder: transformer
encoder_conf:
    output_size: 256
    attention_heads: 4
    linear_units: 1024
    num_blocks: 6
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    attention_dropout_rate: 0.1
    input_layer: conv2d2
    normalize_before: true

decoder: transformer
decoder_conf:
    attention_heads: 4
    linear_units: 2048
    num_blocks: 4
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    self_attention_dropout_rate: 0.1
    src_attention_dropout_rate: 0.1

model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1
    length_normalized_loss: false
    extract_feats_in_collect_stats: false

seed: 2022
log_interval: 400
num_att_plot: 0
num_workers: 4
sort_in_batch: descending
sort_batch: descending
batch_type: numel
batch_bins: 12000000
accum_grad: 4
max_epoch: 160
patience: none
init: none
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 4

use_amp: true
cudnn_deterministic: false
cudnn_benchmark: false


optim: adam
optim_conf:
    lr: 0.008
    weight_decay: 0.001
scheduler: warmuplr
scheduler_conf:
    warmup_steps: 1000


specaug: specaug
specaug_conf:
    apply_time_warp: true
    time_warp_window: 5
    time_warp_mode: bicubic
    apply_freq_mask: true
    freq_mask_width_range:
    - 0
    - 27
    num_freq_mask: 2
    apply_time_mask: true
    time_mask_width_ratio_range:
    - 0.
    - 0.05
    num_time_mask: 5
```

**Comment by ftshijt** [comment]
Thanks for the update! I feel the results shall be sufficient now. As I go through the paper, I see that the data is designed for evaluation purposes only. That is saying that they do allow additional training data involved. Therefore, it would be quite reasonable to have the current results. 

Thanks for your great contribution so far. Please finish the PR by updating the config to the latest~

**Comment by ftshijt** [comment]
Thanks for your contribution!
